utils/filecache/cache.go

```diff
@@ -0,0 +1,200 @@
+	"context"
+	"errors"
+	"sync/atomic"
+	"time"
+
+	"github.com/ARM-software/golang-utils/utils/commonerrors"
+	fs "github.com/ARM-software/golang-utils/utils/filesystem"
+	"github.com/ARM-software/golang-utils/utils/parallelisation"
+)
+
+type Cache struct {
+	entries       *entryMap
+	entriesLM     *lockMap
+	entryProvider IEntryProvider
+	fs            fs.FS
+	cfg           *FileCacheConfig
+	cancelStore   *parallelisation.CancelFunctionStore
+	closed        atomic.Bool
+}
+
+func (c *Cache) gc(ctx context.Context, _ time.Time) {
+	c.entries.Range(func(key string, entry ICacheEntry) {
+		if entry.IsExpired() && c.entriesLM.TryLock(key) {
+			if err := entry.Delete(ctx, c.fs); err == nil {
+				c.entries.Delete(key)
+				c.entriesLM.Unlock(key)
+			}
+		}
+	})
+}
+
+func (c *Cache) checkIfClosed() error {
+	if c.closed.Load() {
+		return commonerrors.New(commonerrors.ErrConflict, "cache is closed")
+	}
+
+	return nil
+}
+
+func (c *Cache) Has(ctx context.Context, key string) (bool, error) {
+	if err := c.checkIfClosed(); err != nil {
+		return false, err
+	}
+
+	return c.entries.Exists(key), nil
+}
+
+func (c *Cache) Store(ctx context.Context, key string) error {
+	return c.StoreWithTTL(ctx, key, c.cfg.TTL)
+}
+
+func (c *Cache) StoreWithTTL(ctx context.Context, key string, ttl time.Duration) error {
+	if err := c.checkIfClosed(); err != nil {
+		return err
+	}
+
+	if ttl == 0 {
+		return commonerrors.New(commonerrors.ErrInvalid, "TTL for an entry can't be zero")
+	}
+
+	if c.entries.Exists(key) {
+		return commonerrors.Newf(commonerrors.ErrExists, "cache entry %s already exists", key)
+	}
+
+	entryPath, err := c.entryProvider.FetchEntry(ctx, key)
+	if err != nil {
+		return err
+	}
+
+	c.entries.Store(key, NewCacheEntry(entryPath, ttl))
+	c.entriesLM.Store(key)
+
+	return nil
+}
+
+func (c *Cache) Fetch(ctx context.Context, key string, destFilesystem fs.FS, destPath string) error {
+	if err := c.checkIfClosed(); err != nil {
+		return err
+	}
+
+	c.entriesLM.Lock(key)
+	defer c.entriesLM.Unlock(key)
+
+	if !c.entries.Exists(key) {
+		return commonerrors.Newf(commonerrors.ErrNotFound, "cache entry for %s not found", key)
+	}
+
+	cpCtx, stop := context.WithCancel(ctx)
+	c.cancelStore.RegisterCancelFunction(stop)
+
+	entry := c.entries.Load(key)
+	if err := entry.Copy(cpCtx, c.fs, destFilesystem, destPath); err != nil {
+		return err
+	}
+
+	entry.ExtendLifetime()
+
+	return nil
+}
+
+func (c *Cache) Evict(ctx context.Context, key string) error {
+	if err := c.checkIfClosed(); err != nil {
+		return err
+	}
+
+	c.entriesLM.Lock(key)
+	defer c.entriesLM.Unlock(key)
+
+	if c.entries.Exists(key) {
+
+		entry := c.entries.Load(key)
+		if err := entry.Delete(ctx, c.fs); err != nil {
+			return err
+		}
+
+		c.entries.Delete(key)
+	}
+
+	return nil
+}
+
+func (c *Cache) Close() error {
+	if err := c.checkIfClosed(); err != nil {
+		return nil
+	}
+
+	c.closed.Store(true)
+	c.entriesLM.Clear()
+	c.cancelStore.Cancel()
+
+	ctx := context.Background()
```

Comment:
this close function may not be cancellable. since you use context.Background here, we cannot every cancel it.

I know it may not be something you don't want to cancel since it is cleaning up, but when we run things in kubernetes etc. they use signals to shutdown pods etc.

If a pod was being deleted then this cleanup would block and cause the pod to be stuck terminating which might cause issues, especially since if we were removing a pod or something we would be getting rid of the filesystem data anyway